app/actions/v3/service_instance_delete.rb

```diff
@@ -14,20 +16,38 @@ def delete(service_instance)
 
         cannot_delete_shared_instances! if service_instance.shared?
 
-        case service_instance
-        when ManagedServiceInstance
-          raise NotImplementedError
-        end
+        lock = DeleterLock.new(service_instance)
+        lock.lock!
+
+        job = case service_instance
+              when ManagedServiceInstance
+                asynchronous_destroy(service_instance)
+              when UserProvidedServiceInstance
+                synchronous_destroy(service_instance, lock)
+              end
 
-        service_instance.db.transaction do
-          service_instance.lock!
-          service_instance.destroy
-          service_event_repository.record_user_provided_service_instance_event(:delete, service_instance, {})
-        end
+        job
       end
 
       private
 
+      def synchronous_destroy(service_instance, lock)
+        lock.unlock_and_destroy!
+        service_event_repository.record_user_provided_service_instance_event(:delete, service_instance, {})
+        nil
+      end
+
+      def asynchronous_destroy(service_instance)
+        delete_job = V3::DeleteServiceInstanceJob.new(
+          service_instance.guid,
+          :deprovision,
```

Comment:
It feels a bit like we are exposing the details of the Job by providing the method we are going to send to the broker client within the action.  It also feels a bit weird since the method that will be called on the client is being put into the database through the job queueing.

I think you hinted at a potential refactor across all these jobs, so it makes sense to keep this for now and maybe after the refactoring there might be an obvious way to expose this detail outside the job?

Reply:
Yeah, this is pretty much a work in progress at this point and not quite there yet. I'll keep your point in mind when consolidating the jobs 🤔